File: src/assets.ts
import { constants } from "node:fs";
import { open, realpath } from "node:fs/promises";
import { extname, isAbsolute, relative, resolve, sep } from "node:path";

import { XMLParser, XMLValidator } from "fast-xml-parser";

import { ExpectedError, type ErrorSource } from "./errors.ts";

const SVG_NAMESPACE = "http://www.w3.org/2000/svg";
const SAFE_FRAGMENT_ID = /^[A-Za-z_][A-Za-z0-9_.:-]*$/;
const ACTIVE_ELEMENTS = new Set([
  "a",
  "audio",
  "discard",
  "embed",
  "foreignobject",
  "iframe",
  "image",
  "mpath",
  "object",
  "script",
  "set",
  "style",
  "use",
  "video",
]);

const MIME_BY_EXTENSION: Readonly<Record<string, string>> = Object.freeze({
  ".gif": "image/gif",
  ".jpeg": "image/jpeg",
  ".jpg": "image/jpeg",
  ".png": "image/png",
  ".svg": "image/svg+xml",
  ".webp": "image/webp",
});

interface OrderedXmlNode {
  readonly [name: string]: unknown;
  readonly ":@"?: Readonly<Record<string, unknown>>;
}

export interface ImageAssetContext extends ErrorSource {
  readonly assetBase: string;
}

export interface EmbeddedImageAsset {
  readonly mimeType: string;
  readonly dataUri: string;
  readonly canonicalPath: string;
}

function assetFailure(context: ImageAssetContext, message: string): never {
  throw new ExpectedError(message, context);
}

function hasNodeCode(error: unknown, ...codes: string[]): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    codes.includes(String(error.code))
  );
}

function unsafeSvg(context: ImageAssetContext, message: string): never {
  throw new ExpectedError(message, context);
}

function isContained(base: string, target: string): boolean {
  const fromBase = relative(base, target);
  return (
    fromBase === "" ||
    (!isAbsolute(fromBase) && fromBase !== ".." && !fromBase.startsWith(`..${sep}`))
  );
}

function decodeLocalPath(value: string, context: ImageAssetContext): string {
  const pathPart = value.split(/[?#]/u, 1)[0] ?? "";
  if (pathPart === "" || pathPart.startsWith("/") || pathPart.startsWith("\\")) {
    assetFailure(context, "Image path must be a relative local path.");
  }
  if (pathPart.includes("\\")) {
    assetFailure(context, "Image path contains an ambiguous path separator.");
  }

  let decoded: string;
  try {
    decoded = decodeURIComponent(pathPart);
  } catch {
    assetFailure(context, "Image path has invalid percent encoding.");
  }
  if (decoded === "" || decoded.includes("\0") || decoded.includes("\\") || isAbsolute(decoded)) {
    assetFailure(context, "Image path is invalid.");
  }
  return decoded;
}

function hasPrefix(bytes: Uint8Array, signature: readonly number[]): boolean {
  return signature.every((byte, index) => bytes[index] === byte);
}

function hasTrustedSignature(bytes: Uint8Array, extension: string): boolean {
  switch (extension) {
    case ".png":
      return hasPrefix(bytes, [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
    case ".jpg":
    case ".jpeg":
      return hasPrefix(bytes, [0xff, 0xd8, 0xff]);
    case ".gif":
      return (
        new TextDecoder().decode(bytes.subarray(0, 6)) === "GIF87a" ||
        new TextDecoder().decode(bytes.subarray(0, 6)) === "GIF89a"
      );
    case ".webp":
      return (
        new TextDecoder().decode(bytes.subarray(0, 4)) === "RIFF" &&
        new TextDecoder().decode(bytes.subarray(8, 12)) === "WEBP"
      );
    case ".svg":
      return true;
    default:
      return false;
  }
}

function checkCssValue(value: string, context: ImageAssetContext): void {
  if (
    /@import\b/iu.test(value) ||
    /expression\s*\(/iu.test(value) ||
    (value.includes("\\") && value.includes("("))
  ) {
    unsafeSvg(context, "SVG contains unsafe CSS.");
  }

  const urlStart = /url\s*\(/giu;
  while (urlStart.exec(value) !== null) {
    const close = value.indexOf(")", urlStart.lastIndex);
    if (close === -1) unsafeSvg(context, "SVG contains a malformed CSS URL.");
    const target = value.slice(urlStart.lastIndex, close).trim();
    if (!target.startsWith("#") || !SAFE_FRAGMENT_ID.test(target.slice(1))) {
      unsafeSvg(context, "SVG contains a non-fragment CSS URL.");
    }
    urlStart.lastIndex = close + 1;
  }
}

function nodeEntries(node: OrderedXmlNode): Array<readonly [string, unknown]> {
  return Object.entries(node).filter(([name]) => name !== ":@");
}

function inspectSvgNode(node: OrderedXmlNode, context: ImageAssetContext, isRoot = false): void {
  const attributes = node[":@"];
  if (attributes !== undefined) {
    for (const [rawName, rawValue] of Object.entries(attributes)) {
      const name = rawName.toLowerCase();
      if (name.startsWith("on") || name === "href" || name === "xlink:href" || name === "src") {
        unsafeSvg(context, `SVG contains unsafe attribute '${rawName}'.`);
      }
      if (
        rawName.includes(":") ||
        name.startsWith("xmlns:") ||
        (name === "xmlns" && (!isRoot || rawValue !== SVG_NAMESPACE))
      ) {
        unsafeSvg(context, `SVG contains ambiguous namespaced attribute '${rawName}'.`);
      }
      if (typeof rawValue !== "string") {
        unsafeSvg(context, `SVG attribute '${rawName}' does not have a string value.`);
      }
      checkCssValue(rawValue, context);
    }
  }

  for (const [rawName, rawChildren] of nodeEntries(node)) {
    if (rawName === "#text" || rawName === "#cdata") {
      const textNodes = Array.isArray(rawChildren) ? rawChildren : [];
      for (const textNode of textNodes) {
        if (typeof textNode !== "object" || textNode === null) continue;
        const value = (textNode as { readonly "#text"?: unknown })["#text"];
        if (typeof value === "string") checkCssValue(value, context);
      }
      continue;
    }
    if (rawName === "#comment") continue;
    if (rawName.startsWith("?")) unsafeSvg(context, "SVG contains a processing instruction.");
    if (rawName.includes(":")) unsafeSvg(context, "SVG contains an ambiguous namespaced element.");

    const name = rawName.toLowerCase();
    if (ACTIVE_ELEMENTS.has(name) || name.startsWith("animate")) {
      unsafeSvg(context, `SVG contains active element '${rawName}'.`);
    }
    if (!Array.isArray(rawChildren)) unsafeSvg(context, "SVG parse structure is ambiguous.");
    for (const child of rawChildren) {
      if (typeof child !== "object" || child === null || Array.isArray(child)) {
        unsafeSvg(context, "SVG parse structure is ambiguous.");
      }
      inspectSvgNode(child as OrderedXmlNode, context);
    }
  }
}

/** Validate authored SVG structurally while retaining its original bytes for image-mode embedding. */
export function validateAuthoredSvg(bytes: Uint8Array, context: ImageAssetContext): void {
  let svg: string;
  try {
    svg = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    unsafeSvg(context, "SVG is not valid UTF-8.");
  }

  if (/<!\s*(?:DOCTYPE|ENTITY)\b/iu.test(svg) || /&[A-Za-z_:][A-Za-z0-9_.:-]*;/u.test(svg)) {
    unsafeSvg(context, "SVG declarations and entities are not allowed.");
  }

  const withoutBom = svg.startsWith("\uFEFF") ? svg.slice(1) : svg;
  const processingInstructions = [...withoutBom.matchAll(/<\?[\s\S]*?\?>/gu)];
  if (processingInstructions.length > 1) {
    unsafeSvg(context, "SVG contains a processing instruction.");
  }
  if (processingInstructions.length === 1) {
    const instruction = processingInstructions[0]!;
    if (instruction.index !== 0 || !/^<\?xml(?:\s[^?]*)?\?>$/u.test(instruction[0])) {
      unsafeSvg(context, "SVG contains a non-leading processing instruction.");
    }
  }

  const validation = XMLValidator.validate(svg, { allowBooleanAttributes: false });
  if (validation !== true) unsafeSvg(context, "SVG is malformed XML.");

  let parsed: unknown;
  try {
    parsed = new XMLParser({
      preserveOrder: true,
      ignoreAttributes: false,
      attributeNamePrefix: "",
      allowBooleanAttributes: false,
      parseTagValue: false,
      parseAttributeValue: false,
      trimValues: false,
      processEntities: false,
      ignoreDeclaration: false,
      ignorePiTags: false,
      commentPropName: "#comment",
      cdataPropName: "#cdata",
      removeNSPrefix: false,
    }).parse(svg);
  } catch {
    unsafeSvg(context, "SVG could not be parsed safely.");
  }
  if (!Array.isArray(parsed)) unsafeSvg(context, "SVG parse structure is ambiguous.");

  const roots: OrderedXmlNode[] = [];
  for (const item of parsed) {
    if (typeof item !== "object" || item === null || Array.isArray(item)) {
      unsafeSvg(context, "SVG parse structure is ambiguous.");
    }
    const record = item as OrderedXmlNode;
    for (const [name] of nodeEntries(record)) {
      if (name === "?xml" || name === "#comment" || name === "#text") continue;
      if (name === "#cdata" || name.startsWith("?")) {
        unsafeSvg(context, "SVG contains content outside its root element.");
      }
      roots.push(record);
    }
  }
  if (roots.length !== 1 || !Object.hasOwn(roots[0]!, "svg")) {
    unsafeSvg(context, "SVG must contain exactly one unprefixed svg root.");
  }

  const root = roots[0]!;
  const attributes = root[":@"] ?? {};
  if (attributes.xmlns !== SVG_NAMESPACE) {
    unsafeSvg(context, "SVG root must declare the SVG namespace unambiguously.");
  }
  if (Object.keys(attributes).some((name) => name.startsWith("xmlns:"))) {
    unsafeSvg(context, "SVG contains an ambiguous namespace declaration.");
  }
  inspectSvgNode(root, context, true);
}

/** Resolve, contain, open, verify, and encode one local image candidate. */
export async function embedLocalImageAsset(
  sourceUrl: string,
  context: ImageAssetContext,
): Promise<EmbeddedImageAsset> {
  const decodedPath = decodeLocalPath(sourceUrl, context);

  let canonicalBase: string;
  try {
    canonicalBase = await realpath(resolve(context.assetBase));
  } catch {
    assetFailure(context, "Image asset base could not be resolved.");
  }

  const lexicalTarget = resolve(canonicalBase, decodedPath);
  if (!isContained(canonicalBase, lexicalTarget)) {
    assetFailure(context, "Image path escapes the source directory.");
  }

  let canonicalTarget: string;
  try {
    canonicalTarget = await realpath(lexicalTarget);
  } catch (error) {
    if (hasNodeCode(error, "ENOENT", "ENOTDIR")) {
      assetFailure(context, "Image asset was not found.");
    }
    assetFailure(context, "Image asset could not be read or resolved.");
  }
  if (!isContained(canonicalBase, canonicalTarget)) {
    assetFailure(context, "Image asset resolves outside the source directory.");
  }

  const extension = extname(canonicalTarget).toLowerCase();
  const mimeType = MIME_BY_EXTENSION[extension];
  if (mimeType === undefined) assetFailure(context, "Image type is not supported.");

  let handle: Awaited<ReturnType<typeof open>> | undefined;
  let bytes: Uint8Array;
  try {
    handle = await open(canonicalTarget, constants.O_RDONLY | constants.O_NOFOLLOW);
    const before = await handle.stat();
    if (!before.isFile()) assetFailure(context, "Image asset is not a regular file.");
    bytes = new Uint8Array(await handle.readFile());
    const after = await handle.stat();
    if (
      !after.isFile() ||
      before.dev !== after.dev ||
      before.ino !== after.ino ||
      before.size !== after.size ||
      bytes.byteLength !== after.size
    ) {
      assetFailure(context, "Image asset changed while it was being read.");
    }
  } catch (error) {
    if (error instanceof ExpectedError) throw error;
    assetFailure(context, "Image asset could not be read.");
  } finally {
    await handle?.close().catch(() => undefined);
  }

  if (!hasTrustedSignature(bytes, extension)) {
    assetFailure(context, "Image contents do not match the supported image type.");
  }
  if (extension === ".svg") validateAuthoredSvg(bytes, context);

  return Object.freeze({
    mimeType,
    canonicalPath: canonicalTarget,
    dataUri: `data:${mimeType};base64,${Buffer.from(bytes).toString("base64")}`,
  });
}

File: src/assets.rs
use std::fs;
use std::path::{Path, PathBuf};

use base64::Engine as _;
use base64::engine::general_purpose::STANDARD;
use percent_encoding::percent_decode_str;
use url::Url;

use crate::AppError;

pub(crate) fn resolve_remote_image(
    raw: &str,
    source_base: &Url,
    source_label: &str,
    line: usize,
    column: usize,
) -> Result<String, AppError> {
    let value = raw.trim();
    let context = format!("{source_label}:{line}:{column}");
    if value.is_empty() {
        return Err(AppError::labeled("Image URL is empty.", context));
    }
    if value.starts_with("//") {
        return Err(AppError::labeled(
            "Protocol-relative image URLs are not allowed.",
            context,
        ));
    }
    if value.contains('\\') {
        return Err(AppError::labeled(
            "Image URL is unsafe or invalid.",
            context,
        ));
    }

    if let Ok(url) = Url::parse(value) {
        return if matches!(url.scheme(), "http" | "https") && url.host_str().is_some() {
            Ok(value.to_owned())
        } else {
            Err(AppError::labeled(
                "Image URL scheme is not allowed.",
                context,
            ))
        };
    }

    let joined = source_base
        .join(value)
        .map_err(|_| AppError::labeled("Image URL is unsafe or invalid.", context.clone()))?;
    if matches!(joined.scheme(), "http" | "https") && joined.host_str().is_some() {
        Ok(joined.to_string())
    } else {
        Err(AppError::labeled(
            "Image URL scheme is not allowed.",
            context,
        ))
    }
}

pub(crate) fn resolve_image(
    raw: &str,
    source_base: &Path,
    source_label: &str,
    line: usize,
    column: usize,
) -> Result<String, AppError> {
    let value = raw.trim();
    let context = format!("{source_label}:{line}:{column}");

    if value.is_empty() {
        return Err(AppError::labeled("Image URL is empty.", context));
    }
    if value.starts_with("//") {
        return Err(AppError::labeled(
            "Protocol-relative image URLs are not allowed.",
            context,
        ));
    }

    if let Ok(url) = Url::parse(value) {
        return match url.scheme() {
            "http" | "https" if url.host_str().is_some() => Ok(value.to_owned()),
            "http" | "https" => Err(AppError::labeled("Remote image URL is invalid.", context)),
            _ => Err(AppError::labeled(
                "Image URL scheme is not allowed.",
                context,
            )),
        };
    }

    let path_part = value.split(['?', '#']).next().unwrap_or_default();
    if path_part.is_empty()
        || path_part.starts_with(['/', '\\'])
        || path_part.contains('\\')
        || Path::new(path_part).is_absolute()
    {
        return Err(AppError::labeled(
            "Image path must be a relative local path.",
            context,
        ));
    }
    validate_percent_encoding(path_part)
        .map_err(|message| AppError::labeled(message, context.clone()))?;
    let decoded = percent_decode_str(path_part)
        .decode_utf8()
        .map_err(|_| AppError::labeled("Image path is not valid UTF-8.", context.clone()))?;
    if decoded.is_empty()
        || decoded.contains('\0')
        || decoded.contains('\\')
        || Path::new(decoded.as_ref()).is_absolute()
    {
        return Err(AppError::labeled("Image path is invalid.", context));
    }

    let canonical_base = fs::canonicalize(source_base).map_err(|_| {
        AppError::labeled(
            "Image source directory could not be resolved.",
            context.clone(),
        )
    })?;
    let target = canonicalize_target(&canonical_base, decoded.as_ref(), &context)?;
    let metadata = fs::metadata(&target)
        .map_err(|_| AppError::labeled("Image asset could not be read.", context.clone()))?;
    if !metadata.is_file() {
        return Err(AppError::labeled(
            "Image asset is not a regular file.",
            context,
        ));
    }

    let mime = mime_for_path(&target)
        .ok_or_else(|| AppError::labeled("Image type is not supported.", context.clone()))?;
    let bytes = fs::read(&target)
        .map_err(|_| AppError::labeled("Image asset could not be read.", context))?;
    Ok(format!("data:{mime};base64,{}", STANDARD.encode(bytes)))
}

fn canonicalize_target(base: &Path, relative: &str, context: &str) -> Result<PathBuf, AppError> {
    let lexical = base.join(relative);
    if !lexical.starts_with(base) {
        return Err(AppError::labeled(
            "Image path escapes the source directory.",
            context,
        ));
    }
    let canonical = fs::canonicalize(&lexical).map_err(|error| {
        let message = if error.kind() == std::io::ErrorKind::NotFound {
            "Image asset was not found."
        } else {
            "Image asset could not be resolved."
        };
        AppError::labeled(message, context)
    })?;
    if !canonical.starts_with(base) {
        return Err(AppError::labeled(
            "Image asset resolves outside the source directory.",
            context,
        ));
    }
    Ok(canonical)
}

fn validate_percent_encoding(value: &str) -> Result<(), &'static str> {
    let bytes = value.as_bytes();
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'%' {
            if index + 2 >= bytes.len()
                || !bytes[index + 1].is_ascii_hexdigit()
                || !bytes[index + 2].is_ascii_hexdigit()
            {
                return Err("Image path has invalid percent encoding.");
            }
            index += 3;
        } else {
            index += 1;
        }
    }
    Ok(())
}

fn mime_for_path(path: &Path) -> Option<&'static str> {
    match path
        .extension()
        .and_then(|extension| extension.to_str())?
        .to_ascii_lowercase()
        .as_str()
    {
        "png" => Some("image/png"),
        "jpg" | "jpeg" => Some("image/jpeg"),
        "gif" => Some("image/gif"),
        "webp" => Some("image/webp"),
        "svg" => Some("image/svg+xml"),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use tempfile::tempdir;

    use super::*;

    #[test]
    fn resolves_remote_images_without_local_access_and_rejects_unsafe_forms() {
        let base = Url::parse("https://example.test/docs/guide.md?q=old#part").unwrap();
        for (raw, expected) in [
            ("image.png", "https://example.test/docs/image.png"),
            ("/assets/root.png", "https://example.test/assets/root.png"),
            ("?raw=1", "https://example.test/docs/guide.md?raw=1"),
            (
                "#preview",
                "https://example.test/docs/guide.md?q=old#preview",
            ),
            (
                "https://cdn.example.test/a.png?q=1",
                "https://cdn.example.test/a.png?q=1",
            ),
        ] {
            assert_eq!(
                resolve_remote_image(raw, &base, "remote", 2, 3).unwrap(),
                expected
            );
        }
        for unsafe_value in [
            "//example.test/a.png",
            "file:///tmp/a.png",
            "data:image/png;base64,AAAA",
            "javascript:alert(1)",
            "..\\local.png",
            "",
        ] {
            let error = resolve_remote_image(unsafe_value, &base, "remote", 2, 3).unwrap_err();
            assert!(error.to_string().starts_with("remote:2:3:"), "{error}");
        }
    }

    #[test]
    fn embeds_supported_extensions_without_signature_validation() {
        let directory = tempdir().unwrap();
        for (name, mime) in [
            ("a.png", "image/png"),
            ("a.jpg", "image/jpeg"),
            ("a.gif", "image/gif"),
            ("a.webp", "image/webp"),
            ("a.svg", "image/svg+xml"),
        ] {
            fs::write(directory.path().join(name), b"authored bytes").unwrap();
            let uri = resolve_image(name, directory.path(), "doc.md", 2, 1).unwrap();
            assert_eq!(
                uri,
                format!("data:{mime};base64,{}", STANDARD.encode(b"authored bytes"))
            );
        }
    }

    #[test]
    fn decodes_nested_paths_and_discards_query_and_fragment() {
        let directory = tempdir().unwrap();
        let nested = directory.path().join("Unicode ü space");
        fs::create_dir(&nested).unwrap();
        fs::write(nested.join("tiny image.png"), b"png").unwrap();
        let uri = resolve_image(
            "Unicode%20%C3%BC%20space/tiny%20image.png?download=1#preview",
            directory.path(),
            "stdin",
            1,
            1,
        )
        .unwrap();
        assert_eq!(uri, "data:image/png;base64,cG5n");
    }

    #[test]
    fn preserves_remote_urls_and_rejects_unsafe_or_missing_local_paths() {
        let directory = tempdir().unwrap();
        let remote = "https://images.example.test/a.png?q=1#view";
        assert_eq!(
            resolve_image(remote, directory.path(), "doc.md", 1, 1).unwrap(),
            remote
        );

        for value in [
            "data:image/png;base64,AAAA",
            "file:///tmp/a.png",
            "//example.test/a.png",
            "/tmp/a.png",
            "../outside.png",
            "missing.png",
            "bad%ZZ.png",
        ] {
            let error = resolve_image(value, directory.path(), "doc.md", 7, 3).unwrap_err();
            assert!(error.to_string().starts_with("doc.md:7:3:"), "{error}");
        }
    }

    #[cfg(unix)]
    #[test]
    fn rejects_symlinks_that_escape_the_canonical_source_base() {
        use std::os::unix::fs::symlink;

        let root = tempdir().unwrap();
        let base = root.path().join("base");
        fs::create_dir(&base).unwrap();
        let outside = root.path().join("outside.png");
        fs::write(&outside, b"png").unwrap();
        symlink(&outside, base.join("escape.png")).unwrap();

        let error = resolve_image("escape.png", &base, "doc.md", 1, 1).unwrap_err();
        assert!(error.to_string().contains("outside the source directory"));
    }
}
